src/plugins/discover/public/utils/get_sharing_data.ts

```diff
@@ -72,6 +77,23 @@ export async function getSharingData(
         searchSource.setField('filter', filter);
       }
 
+      /*
+       * For downstream querying performance, the searchSource object must have fields set.
+       * Otherwise, the requests will ask for all fields, even if only a few are really needed.
+       * Discover does not set fields, since having all fields is needed for the UI.
+       */
+      // 1. Inspect the searchsource to find how fields are configured (fields API or "raw data")
+      const fieldConfigs: Pick<SearchSourceFields, 'fields' | 'fieldsFromSource'> = {
+        fields: searchSource.getField('fields'),
+        fieldsFromSource: searchSource.getField('fieldsFromSource'),
+      };
+      const fieldsConfig = fieldConfigs.fields ? 'fields' : 'fieldsFromSource';
+
+      // 2. explicitly set the fields into the searchsource object
+      if (fieldsConfig === 'fields' && columns.length > 0) {
+        searchSource.setField('fields', columns);
+      }
+
```

Comment:
It seems like it can be simplified. Or am I missing something?

```suggestion
      if (searchSource.getField('fields') && columns.length) {
        searchSource.setField('fields', columns);
      }

```

Reply:
Why not using uiSettings config for that like
`const useFieldsApi = !uiSettings.get(SEARCH_FIELDS_FROM_SOURCE)`